src/main/java/org/folio/fqm/config/EntityTypePermissionsAspect.java

```diff
@@ -0,0 +1,146 @@
+package org.folio.fqm.config;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.annotation.Around;
+import org.aspectj.lang.annotation.Aspect;
+import org.aspectj.lang.reflect.MethodSignature;
+import org.folio.fqm.annotation.EntityTypePermissionsRequired;
+import org.folio.fqm.exception.EntityTypeNotFoundException;
+import org.folio.fqm.exception.MissingPermissionsException;
+import org.folio.fqm.repository.EntityTypeRepository;
+import org.folio.fqm.repository.QueryRepository;
+import org.folio.querytool.domain.dto.ContentsRequest;
+import org.folio.querytool.domain.dto.EntityType;
+import org.folio.querytool.domain.dto.SubmitQuery;
+import org.folio.spring.FolioExecutionContext;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.*;
+import java.util.function.Function;
+
+@Aspect
+@Component
+@RequiredArgsConstructor(onConstructor_ = @Autowired)
+@Log4j2
+public class EntityTypePermissionsAspect {
+
+  private final EntityTypeRepository entityTypeRepository;
+  private final FolioExecutionContext folioExecutionContext;
+  private final QueryRepository queryRepository;
+
+  private final Map<MethodSignature, Integer> indexCache = new HashMap<>();
+
+  /**
+   * Handle methods that accept an entity type or query ID
+   */
+  @Around("@annotation(org.folio.fqm.annotation.EntityTypePermissionsRequired) && execution(* *(.., java.util.UUID, ..))")
```

Comment:
This magic annotation tells Spring/AspectJ to look for methods annotated with `EntityTypePermissionsRequired` that contain a parameter of type `UUID`. The other methods here follow the same pattern

Reply:
scary, but pretty cool!